MovingBlocks/Terasology
# Telemetry API

### Contains

- A Telemetry API that could be used in the module. The test module is here: https://github.com/GabrielXia/TelemetryApiTest 
    - The new telemetry in module will also be shown in the `Metrics Menu` (after a game launch of cause, since in the `main menu` you have not chosen the module)
   - The option for the module field is also valid
- Follow-up to comments in #3016: mark ip address and timezone field as "anonyme"
- Follow-up to #3056: check and update the map which notes the user's telemetry options when game initialise() so the value could be null, that npe could not happen 
- Rewrite the `Metrics.class` which store all the `Metric.class` instances. Now if the new `Metric` instance is created using the constructor in parent class `Metric.class`, it will be added to `Metrics` automatically. 

### How to test

Fetch the test module in https://github.com/GabrielXia/TelemetryApiTest , enable the telemetry in the `Metrics Menu`, select the module and load the game. The `TelemetryApiTestMetric` will be sent to the server and you can see the fields in the `Metrics Menu` in game. You can see the metric sent in http://utility.terasology.org:5601/
